### PR TITLE
Integrate softmax cpr to the latest RecBole version

### PR DESCRIPTION
**Pull request by ken77921**
The code is modified from https://github.com/iesl/softmax_CPR_recommend

### CODE REVIEW

File: recbole/model/sequential_recommender/gru4reccpr.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+            all_memory = gru_output
+            all_memory_values = all_memory
+            bsz, seq_len, hsz = all_memory.size()
+
+            all_memory_U = self.dropout(self.Ue(all_memory))
+            all_memory_U = all_memory_U.unsqueeze(2)
+            all_memory_U = all_memory_U.expand(bsz, seq_len, all_memory.size(1), hsz)
+
+            all_memory_W = self.dropout(self.We(all_memory))
+            all_memory_W = all_memory_U.unsqueeze(1)
```
Comment (BishopLiu) [inline recbole/model/sequential_recommender/gru4reccpr.py:172]: Here should be `all_memory_W`

File: recbole/model/sequential_recommender/gru4reccpr.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+            all_memory_W = all_memory_U.expand(bsz, all_memory.size(1), seq_len, hsz)
```
Comment (BishopLiu) [inline recbole/model/sequential_recommender/gru4reccpr.py:173]: `all_memory_U` here should be replaced by `all_memory_W`

File: recbole/model/sequential_recommender/gru4reccpr.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+                        facet_lm_logits_arr[i].scatter_add_(2, candidate_token_ids, logit_hidden_reranker_topn) #(bsz, seq_len, vocab_size) <- (bsz, seq_len, topk) x (bsz, seq_len, topk)
+                    else:
+                        facet_lm_logits_arr[i].scatter_(2, candidate_token_ids, logit_hidden_reranker_topn) #(bsz, seq_len, vocab_size) <- (bsz, seq_len, topk) x (bsz, seq_len, topk)
+
+            for i in range(self.n_facet_context):
+                bsz, seq_len_1, hidden_size = projected_emb_arr[i].size()
+                bsz, seq_len_2 = item_seq.size()
+                logit_hidden_context = (projected_emb_arr[self.n_facet + self.n_facet_reranker*len(self.reranker_CAN_NUM) + i].unsqueeze(dim=2).expand(-1,-1,seq_len_2,-1) * test_item_emb[item_seq, :].unsqueeze(dim=1).expand(-1,seq_len_1,-1,-1)  ).sum(dim=-1)
+                if test_item_bias is not None:
+                    logit_hidden_reranker_topn += test_item_bias[item_seq].unsqueeze(dim=1).expand(-1,seq_len_1,-1)
```
Comment (BishopLiu) [inline recbole/model/sequential_recommender/gru4reccpr.py:290]: `logit_hidden_reranker_topn` here should be `logit_hidden_context`

**Comment by BishopLiu** [comment]
@ken77921 Nice job! I've pointed out some minor mistakes in the code. Besides, you can add your personal information in the header of the model files. And it would be highly appreciated if you can complete the comments of the new parameters in the two yaml files so that the users can understand their meaning quickly. Last but not least, the test functions for these two models need to be added to the file `tests/model/test_model_auto.py`.

**Comment by ken77921** [comment]
@BishopLiu Thanks for the corrections. I am not very sure how to test `tests/model/test_model_auto.py`. Please let me know if I did something wrong in that file.

**Comment by BishopLiu** [comment]
@ken77921 Thank you for your great contribution! I'll merge it.

**Comment by ken77921** [comment]
Thank you for merging the pull request. I have a question. Will the descriptions of GRU4RecCPR and SASRecCPR be added to the documentation (https://recbole.io/docs/user_guide/model_intro.html)?

**Comment by BishopLiu** [comment]
Yes, we will add it during the next version update.
